File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.dart
class Solution {
  bool searchMatrix(List<List<int>> matrix, int target) {
    int lr = 0, rr = matrix.length-1;
    int lc = 0, rc = matrix[0].length-1;
    while(lr <= rr){
        int midRow = (rr + lr) ~/2;
        if(matrix[midRow][0] <= target && (midRow+1 == matrix.length || target < matrix[midRow+1][0])){
            while(lc <= rc){
                int midCol = (rc + lc) ~/2;
                if(matrix[midRow][midCol] == target){
                    return true;
                }
                else if(matrix[midRow][midCol] < target){
                    lc = midCol + 1;
                }
                else {
                    rc = midCol - 1;
                }
            }
            return false;
        }
        else if(matrix[midRow][0] < target){
            lr = midRow + 1;
        }
        else {
            rr = midRow - 1;
        }
    }
    return false;
  }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int low = 0, high = m-1;
        int left = 0, right = n-1;
        while(low <= high){
            int rowMid = (high + low) / 2;
            if(matrix[rowMid][0] <= target && target <= matrix[rowMid][right]){
                while(left <= right){
                    int colMid = (right + left) / 2;
                    if(matrix[rowMid][colMid] == target) return true;
                    else if(matrix[rowMid][colMid] < target){
                        left = colMid + 1;
                    }else {
                        right = colMid - 1;
                    }
                }
            }else if(matrix[rowMid][0] > target){
                high = rowMid - 1;
            }else {
                low = rowMid + 1;
            }
        }
        return false;
    }
}